code/parachain/frame/ibc/src/light_clients.rs

```
@@ -76,19 +84,73 @@ impl tendermint_light_client_verifier::host_functions::HostFunctionsProvider
 
 impl ics07_tendermint::HostFunctionsProvider for HostFunctionsManager {}
 
+pub struct GrandpaHederHashesStorageInstance;
+impl StorageInstance for GrandpaHederHashesStorageInstance {
+	fn pallet_prefix() -> &'static str {
+		"ibc.lightclients.grandpa"
+	}
+
+	const STORAGE_PREFIX: &'static str = "HeaderHashes";
+}
+pub type GrandpaHeaderHashesStorage = StorageValue<
+	GrandpaHederHashesStorageInstance,
+	BoundedVec<H256, ConstU32<GRANDPA_BLOCK_HASHES_CACHE_SIZE>>,
+	ValueQuery,
+>;
+
+pub struct GrandpaHeaderHashesSetStorageInstance;
+impl StorageInstance for GrandpaHeaderHashesSetStorageInstance {
+	fn pallet_prefix() -> &'static str {
+		"ibc.lightclients.grandpa"
+	}
+
+	const STORAGE_PREFIX: &'static str = "HeaderHashesSet";
+}
+pub type GrandpaHeaderHashesSetStorage = StorageValue<
+	GrandpaHeaderHashesSetStorageInstance,
+	BoundedBTreeSet<H256, ConstU32<GRANDPA_BLOCK_HASHES_CACHE_SIZE>>,
+	ValueQuery,
+>;
+
+/// Maximum number of block number to block hash mappings to keep (oldest pruned first).
+const GRANDPA_BLOCK_HASHES_CACHE_SIZE: u32 = 500;
+
 impl grandpa_client_primitives::HostFunctions for HostFunctionsManager {
 	type Header = RelayChainHeader;
 
 	fn ed25519_verify(sig: &ed25519::Signature, msg: &[u8], pub_key: &ed25519::Public) -> bool {
 		pub_key.verify(&msg, sig)
 	}
 
-	fn add_relaychain_headers(header: &Self::Header) {
-		todo!()
+	fn add_relaychain_header_hashes(new_hashes: &[<Self::Header as Header>::Hash]) {
+		if new_hashes.is_empty() {
+			return
+		}
+
+		GrandpaHeaderHashesSetStorage::mutate(|hashes_set| {
+			GrandpaHeaderHashesStorage::mutate(|hashes| {
+				for hash in new_hashes {
+					match hashes.try_push(*hash) {
+						Ok(_) => {},
+						Err(_) => {
+							let old_hash = hashes.remove(0);
+							hashes_set.remove(&old_hash);
+							let _ = hashes.try_push(*hash);
```

why are you ignoring the error?